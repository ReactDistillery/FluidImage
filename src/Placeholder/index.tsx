import * as React from "react";

export interface PlaceholderProps {
  crop: Array<number>;
  backgroundColor?: string;
  fillContainer?: boolean;
  placeholderRef?(ref: Element | null): void;
  children?: any;
}

const styles: React.CSSProperties = {
  padding: {
    display: "inline-block",
    width: "100%",
    position: "relative",
    overflow: "hidden",
    lineHeight: 0
  }
};

const generatePlaceholderStyles = (
  crop: Array<number>,
  fillContainer = false,
  backgroundColor = "#f9f9f9"
) => ({
  ...styles.padding,
  backgroundColor: backgroundColor,
  paddingBottom: fillContainer ? 0 : `${100 * (crop[3] / crop[2])}%`,
  height: fillContainer ? "100%" : 0
});

export default class Placeholder extends React.Component<PlaceholderProps, {}> {
  render() {
    const {
      backgroundColor,
      fillContainer,
      crop,
      placeholderRef,
      children
    } = this.props;
    return (
      <span
        ref={placeholderRef}
        style={generatePlaceholderStyles(crop, fillContainer, backgroundColor)}
      >
        {children}
      </span>
    );
  }
}
