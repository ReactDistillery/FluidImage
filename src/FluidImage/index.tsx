import * as React from "react";
import Measure, {
  ContentRect,
  BoundingRect,
  MeasuredComponentProps
} from "react-measure";

import Placeholder from "../Placeholder";
import Image from "../Image";
import ThumbnailImage from "../ThumbnailImage";
import ErrorBoundary from "../ErrorBoundary";
import { dppx } from '../utils';

export interface FluidImageProps extends React.Props<any> {
  src: string;
  alt?: string;
  title?: string;
  crop: Array<number>;
  pixelRatio?: number;
  quality?: number;
  pattern?: Function;
  thumbnailSize?: number;
  backgroundColor?: string;
  thumbnailPattern?: Function;
  hideThumbnailOnLoad?: boolean;
  fillContainer?: boolean;
}

export interface Pattern {
  src: string;
  width: number;
  height: number;
  crop: Array<number>;
  pixelRatio: number;
  quality: number;
}

export interface FluidImageState {
  hasImageLoaded: boolean;
  hasThumbnailLoaded: boolean;
}

const styles: React.CSSProperties = {
  container: {
    width: "100%",
    position: "absolute"
  }
};

const setThumbnailHasLoaded = (prevState: FluidImageState) => ({
  hasThumbnailLoaded: true
});
const setImageHasLoaded = (prevState: FluidImageState) => ({
  hasImageLoaded: true
});
const setImageUnLoaded = (prevState: FluidImageState) => ({
  hasImageLoaded: false
});
const defaultPattern = ({
  src,
  width,
  crop,
  pixelRatio,
  quality,
  height
}: Pattern): string =>
  `${src}?fit=crop&w=${Math.ceil(width)}&h=${Math.ceil(
    height
  )}&dpr=${pixelRatio}&q=${quality}&rect=${crop.join(",")}`;

export default class FluidImage extends React.Component<
  FluidImageProps,
  FluidImageState
> {
  state = {
    hasImageLoaded: false,
    hasThumbnailLoaded: false
  };

  private handleImageOnLoad = () => this.setState(setImageHasLoaded);
  private handleThumbnailOnLoad = () => this.setState(setThumbnailHasLoaded);
  private handleUpdate = (bounds: BoundingRect | undefined) =>
    this.setState(setImageUnLoaded);

  render() {
    const {
      src,
      alt,
      title,
      crop,
      pixelRatio = dppx(),
      quality = 80,
      pattern = defaultPattern,
      thumbnailSize = 8,
      backgroundColor = "#f9f9f9",
      thumbnailPattern = defaultPattern,
      hideThumbnailOnLoad = false,
      fillContainer = false
    } = this.props;

    const { hasImageLoaded, hasThumbnailLoaded } = this.state;
    return (
      <ErrorBoundary
        fillContainer={fillContainer}
        crop={crop}
        backgroundColor={backgroundColor}
      >
        <Measure
          bounds
          onResize={contentRect => this.handleUpdate(contentRect.bounds)}
        >
          {({ measureRef, contentRect: { bounds } }) => (
            <Placeholder
              placeholderRef={measureRef}
              fillContainer={fillContainer}
              crop={crop}
              backgroundColor={backgroundColor}
            >
              {bounds && (
                <span style={styles.container}>
                  <ThumbnailImage
                    onLoad={this.handleThumbnailOnLoad}
                    src={thumbnailPattern({
                      src,
                      crop,
                      pixelRatio,
                      quality,
                      width: thumbnailSize,
                      height: fillContainer
                        ? bounds.height / bounds.width * thumbnailSize
                        : crop[3] / crop[2] * bounds.width * thumbnailSize
                    })}
                    hasImageLoaded={hasThumbnailLoaded}
                    hideImageOnLoad={hideThumbnailOnLoad}
                    alt={alt}
                    title={title}
                  />
                  {hasThumbnailLoaded ? (
                    <Image
                      onLoad={this.handleImageOnLoad}
                      src={pattern({
                        width: bounds!.width,
                        height: fillContainer
                          ? bounds.height
                          : crop[3] / crop[2] * bounds.width,
                        src,
                        crop,
                        pixelRatio,
                        quality
                      })}
                      title={title}
                      alt={alt}
                      hasImageLoaded={hasImageLoaded}
                    />
                  ) : null}
                </span>
              )}
            </Placeholder>
          )}
        </Measure>
      </ErrorBoundary>
    );
  }
}
