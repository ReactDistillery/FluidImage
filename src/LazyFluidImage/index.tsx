import * as React from "react";
import IntersectionObserver from "react-intersection-observer";
import FluidImage, { FluidImageProps } from "../FluidImage";
import Placeholder from "../Placeholder";

export interface LazyFluidImageState {
  isInView: boolean;
}
export default class LazyFluidImage extends React.Component<
  FluidImageProps,
  LazyFluidImageState
> {
  state = {
    isInView: false
  };

  setIsInView = (isInView: boolean) => this.setState(() => ({ isInView }));
  render() {
    const { isInView } = this.state;
    const { fillContainer, crop, backgroundColor, ...rest } = this.props;
    return (
      <IntersectionObserver
        tag="span"
        onChange={this.setIsInView}
        data-fluid-image
      >
        {isInView ? (
          <FluidImage fillContainer={fillContainer} crop={crop} {...rest} />
        ) : (
          <Placeholder
            fillContainer={fillContainer}
            crop={crop}
            backgroundColor={backgroundColor}
          />
        )}
      </IntersectionObserver>
    );
  }
}
