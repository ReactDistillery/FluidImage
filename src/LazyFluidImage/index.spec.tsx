import React from "react";
import { create } from "react-test-renderer";
import IntersectionObserver from "react-intersection-observer";
import LazyFluidImage from "./";
import Placeholder from "../Placeholder";
import FluidImage from "../FluidImage";

const minimalProps = {
  src: "testImage.jpeg",
  crop: [0, 0, 100, 100]
};

describe("FluidImage - LazyFluidImage", () => {
  it("renders", () => {
    const lazyFluidImage = create(
      <LazyFluidImage {...minimalProps} />
    ).toJSON();
    expect(lazyFluidImage).toMatchSnapshot();
  });
  it("shows `<Placeholder />` when state isInView is false", () => {
    const lazyFluidImage = create(<LazyFluidImage {...minimalProps} />);
    expect(lazyFluidImage.root.instance.state.isInView).toEqual(false);
    expect(lazyFluidImage.root.findByType(Placeholder)).toBeTruthy();
  });
  it("shows `<FluidImage /> when state isInView is true", () => {
    const lazyFluidImage = create(<LazyFluidImage {...minimalProps} />);
    expect(lazyFluidImage.root.instance.state.isInView).toEqual(false);
    lazyFluidImage.root.findByType(IntersectionObserver).props.onChange(true);
    expect(lazyFluidImage.root.findByType(FluidImage)).toBeTruthy();
  });
  it("updates state correctly when it comes in view", () => {
    const lazyFluidImage = create(<LazyFluidImage {...minimalProps} />);
    expect(lazyFluidImage.root.instance.state.isInView).toEqual(false);
    lazyFluidImage.root.findByType(IntersectionObserver).props.onChange(true);
    expect(lazyFluidImage.root.instance.state.isInView).toEqual(true);
  });
});
