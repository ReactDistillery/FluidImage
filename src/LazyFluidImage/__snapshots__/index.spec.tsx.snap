// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FluidImage - LazyFluidImage renders 1`] = `
<span
  data-fluid-image={true}
  onChange={[Function]}
>
  <span
    style={
      Object {
        "backgroundColor": "#f9f9f9",
        "display": "inline-block",
        "height": 0,
        "lineHeight": 0,
        "overflow": "hidden",
        "paddingBottom": "100%",
        "position": "relative",
        "width": "100%",
      }
    }
  />
</span>
`;
